From 3a0ed7f623ee3c60aac5d89289d81524f3b57d84 Mon Sep 17 00:00:00 2001
From: Ken Russell <kbr@chromium.org>
Date: Fri, 28 Aug 2026 19:46:22 -0700
Subject: [PATCH] [M154] GL: add TexSubImage2D workaround for certain texture
 formats.

Original change's description:
> GL: add TexSubImage2D workaround for certain texture formats.
>
> Add a workaround for a bug in the handling of texture uploads of certain
> host-twiddled texture formats; use TexImage2D + TexSubImage2D instead of
> uploading the data in the TexImage2D call. Apply this workaround to
> PowerVR drivers.
>
> (zmo note: this is an rebase for
> https://chromium-review.git.corp.google.com/c/angle/angle/+/8316408.
> Original author kbr asks me to land it while he is OOO. I fail to rebase
> here or upload a local rebased CL to the original issue. All CL credit
> should go to kbr.)
>
> Bug: chromium:553172761
> Change-Id: Ic476f67e9e814d6407a8445ec0d6ed21ec9ea46e
> Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8331247
> Reviewed-by: Geoff Lang <geofflang@chromium.org>
> Auto-Submit: Zhenyao Mo <zmo@chromium.org>
> Commit-Queue: Geoff Lang <geofflang@chromium.org>

(cherry picked from commit 24d17c3981285f68de1a3a6ccfd1d051a3fd30f0)

Bug: chromium:556055661,chromium:553172761
Change-Id: Ic476f67e9e814d6407a8445ec0d6ed21ec9ea46e
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8347507
Reviewed-by: Zhenyao Mo <zmo@chromium.org>
Reviewed-by: Geoff Lang <geofflang@chromium.org>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
(cherry picked from commit 3a0ed7f623ee3c60aac5d89289d81524f3b57d84)

---
 include/platform/autogen/FeaturesGL_autogen.h |  6 ++++
 include/platform/gl_features.json             |  8 +++++
 src/libANGLE/renderer/gl/TextureGL.cpp        | 35 ++++++++++++++++---
 src/libANGLE/renderer/gl/renderergl_utils.cpp |  4 +++
 util/autogen/angle_features_autogen.cpp       |  1 +
 util/autogen/angle_features_autogen.h         |  1 +
 6 files changed, 50 insertions(+), 5 deletions(-)

diff --git a/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h b/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
index ec28d7445a..a72ea6b3cf 100644
--- a/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
+++ b/third_party/angle/include/platform/autogen/FeaturesGL_autogen.h
@@ -272,6 +272,12 @@ struct FeaturesGL : FeatureSetBase
         &members,
     };
 
+    FeatureInfo useTexSubImageForHostTwiddledNpotUploads = {
+        "useTexSubImageForHostTwiddledNpotUploads",
+        FeatureCategory::OpenGLWorkarounds,
+        &members,
+    };
+
     FeatureInfo limitMax3dArrayTextureSizeTo1024 = {
         "limitMax3dArrayTextureSizeTo1024",
         FeatureCategory::OpenGLWorkarounds,
diff --git a/third_party/angle/include/platform/gl_features.json b/third_party/angle/include/platform/gl_features.json
index d2d4671861..fd18975e48 100644
--- a/third_party/angle/include/platform/gl_features.json
+++ b/third_party/angle/include/platform/gl_features.json
@@ -335,6 +335,14 @@
             ],
             "issue": "https://crbug.com/551573368"
         },
+        {
+            "name": "use_tex_sub_image_for_host_twiddled_npot_uploads",
+            "category": "Workarounds",
+            "description": [
+                "For texture formats which would be twiddled on the host CPU, upload via TexSubImage2D instead of TexImage2D."
+            ],
+            "issue": "http://crbug.com/553172761"
+        },
         {
             "name": "limit_max_3d_array_texture_size_to_1024",
             "category": "Workarounds",
diff --git a/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp b/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
index b0bf935c35..d7d875545d 100644
--- a/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
+++ b/third_party/angle/src/libANGLE/renderer/gl/TextureGL.cpp
@@ -98,6 +98,13 @@ LevelInfoGL GetLevelInfo(const angle::FeaturesGL &features,
                        GetEmulatedAlphaChannel(features, originalInternalFormat));
 }
 
+bool IsHostTwiddledFormat(GLenum internalFormat, GLenum format, GLenum type)
+{
+    return (internalFormat == GL_RGB10_A2 && format == GL_RGBA &&
+            type == GL_UNSIGNED_INT_2_10_10_10_REV) ||
+           (internalFormat == GL_SRGB8_ALPHA8 && format == GL_RGBA && type == GL_UNSIGNED_BYTE);
+}
+
 gl::Texture::DirtyBits GetLevelWorkaroundDirtyBits()
 {
     gl::Texture::DirtyBits bits;
@@ -295,11 +302,29 @@ angle::Result TextureGL::setImageHelper(const gl::Context *context,
     if (nativegl::UseTexImage2D(getType()))
     {
         ASSERT(size.depth == 1);
-        ANGLE_GL_TRY_ALWAYS_CHECK(
-            context, functions->texImage2D(nativegl::GetTextureBindingTarget(target),
-                                           static_cast<GLint>(level), texImageFormat.internalFormat,
-                                           size.width, size.height, 0, texImageFormat.format,
-                                           texImageFormat.type, pixels));
+        if (features.useTexSubImageForHostTwiddledNpotUploads.enabled && pixels != nullptr &&
+            (!gl::isPow2(size.width) || !gl::isPow2(size.height)) &&
+            IsHostTwiddledFormat(texImageFormat.internalFormat, texImageFormat.format,
+                                 texImageFormat.type))
+        {
+            ANGLE_GL_TRY_ALWAYS_CHECK(
+                context, functions->texImage2D(
+                             nativegl::GetTextureBindingTarget(target), static_cast<GLint>(level),
+                             texImageFormat.internalFormat, size.width, size.height, 0,
+                             texImageFormat.format, texImageFormat.type, nullptr));
+            ANGLE_GL_TRY(context, functions->texSubImage2D(
+                                      nativegl::GetTextureBindingTarget(target),
+                                      static_cast<GLint>(level), 0, 0, size.width, size.height,
+                                      texImageFormat.format, texImageFormat.type, pixels));
+        }
+        else
+        {
+            ANGLE_GL_TRY_ALWAYS_CHECK(
+                context, functions->texImage2D(
+                             nativegl::GetTextureBindingTarget(target), static_cast<GLint>(level),
+                             texImageFormat.internalFormat, size.width, size.height, 0,
+                             texImageFormat.format, texImageFormat.type, pixels));
+        }
     }
     else
     {
diff --git a/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp b/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
index ff061083f3..87f428e45e 100644
--- a/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
+++ b/third_party/angle/src/libANGLE/renderer/gl/renderergl_utils.cpp
@@ -2448,6 +2448,10 @@ void InitializeFeatures(const FunctionsGL *functions, angle::FeaturesGL *feature
                             IsPowerVR(vendor));
     ANGLE_FEATURE_CONDITION(features, resetTexStorage2DBaseLevel, IsPowerVR(vendor));
 
+    // crbug.com/553172761
+    ANGLE_FEATURE_CONDITION(features, useTexSubImageForHostTwiddledNpotUploads,
+                            false /* IsPowerVR(vendor) */);
+
     ANGLE_FEATURE_CONDITION(features, adjustSrcDstRegionForBlitFramebuffer,
                             IsLinux() || (IsAndroid() && isNvidia) || (IsWindows() && isNvidia) ||
                                 (IsApple() && functions->standard == STANDARD_GL_ES));
diff --git a/third_party/angle/util/autogen/angle_features_autogen.cpp b/third_party/angle/util/autogen/angle_features_autogen.cpp
index a58d3464a5..fc0d07eb90 100644
--- a/third_party/angle/util/autogen/angle_features_autogen.cpp
+++ b/third_party/angle/util/autogen/angle_features_autogen.cpp
@@ -512,6 +512,7 @@ constexpr PackedEnumMap<Feature, const char *> kFeatureNames = {{
     {Feature::UseStencilOpDynamicState, "useStencilOpDynamicState"},
     {Feature::UseStencilTestEnableDynamicState, "useStencilTestEnableDynamicState"},
     {Feature::UseSystemMemoryForConstantBuffers, "useSystemMemoryForConstantBuffers"},
+    {Feature::UseTexSubImageForHostTwiddledNpotUploads, "useTexSubImageForHostTwiddledNpotUploads"},
     {Feature::UseUnusedBlocksWithStandardOrSharedLayout, "useUnusedBlocksWithStandardOrSharedLayout"},
     {Feature::UseVertexInputBindingStrideDynamicState, "useVertexInputBindingStrideDynamicState"},
     {Feature::UseVkEventForBufferBarrier, "useVkEventForBufferBarrier"},
diff --git a/third_party/angle/util/autogen/angle_features_autogen.h b/third_party/angle/util/autogen/angle_features_autogen.h
index dae7430b86..5ddb0103dc 100644
--- a/third_party/angle/util/autogen/angle_features_autogen.h
+++ b/third_party/angle/util/autogen/angle_features_autogen.h
@@ -512,6 +512,7 @@ enum class Feature
     UseStencilOpDynamicState,
     UseStencilTestEnableDynamicState,
     UseSystemMemoryForConstantBuffers,
+    UseTexSubImageForHostTwiddledNpotUploads,
     UseUnusedBlocksWithStandardOrSharedLayout,
     UseVertexInputBindingStrideDynamicState,
     UseVkEventForBufferBarrier,
